.github/workflows/release.yml

```diff
@@ -119,6 +119,11 @@ jobs:
         uses: pypa/cibuildwheel@v2.21.3
         env:
           CIBW_ARCHS: ${{ matrix.arch }}
+          # TEMP don't use automated/isolated build environment, but manually
+          # install build dependencies so we can build with cython 3.1.0a0
+          # once Cython and numpy have official releases that support free threading we can remove this
+          CIBW_BUILD_FRONTEND: "pip; args: --no-build-isolation"
+          CIBW_BEFORE_BUILD: pip install --pre --extra-index-url https://pypi.anaconda.org/scientific-python-nightly-wheels/simple numpy setuptools cython wheel
```

Comment:
Wondering if there is a way to limit this to cython or limit this to the free-threading build (so we still build all the other binaries with released versions)

Reply:
Looking at the numpy and pandas setup, they essentially create one entry in the workflow's matrix per wheel that is being built (instead of letting cibuildwheel generate multiple wheels for the different python versions per platform), and so that way they can control this a bit easier ..

Reply:
I think building a free-threaded wheel with Cython 3.1.0a0 is perfectly fine, because that's the minimum version with any kind of support.

I wouldn't add `>=3.1.0a0` in `pyproject.toml` though, that will give problems for distros and is unnecessary for anything but free-threading. Plus there may be as-yet-undetected regressions, it's an alpha release after all.

Reply:
I thought most distros built with `no-build-isolation` generally so they could bring in their own versions of things. Although I guess they still have to parse that section to find out what the minimum requirement should be on their build dependencies... :+1: I reverted that change now.

Reply:
Yes, the parsing indeed, mostly distros have some tooling for that (e.g., Grayskull for conda-forge).

Reply:
But so then it's still the question how to get cython 3.1.0a0 for just the free-threaded wheel build?

I suppose one other option would be to do some if/else logic in the CIBW_BEFORE_BUILD section. I remember now that I did something like that in https://github.com/apache/arrow/pull/44193/files#diff-1b07285b8a4a94b9f2e56a12d1baf7b7b2f186abc44e5b93f644fd1771f7dc10

Reply:
:+1: Thanks for that link. I think we can put that logic in pretty easily, I added it as a one-liner because I wasn't entirely sure how the line continuations would work for the multi-line if block.